resolwe/flow/managers/listener.py

```diff
@@ -0,0 +1,406 @@
+            )
+
+            d = Data.objects.get(pk=data_id)
+
+            d.process_error.append(exc.message)
+            d.status = Data.STATUS_ERROR
+
+            update_fields = ['process_error', 'status']
+            try:
+                d.save(update_fields=update_fields)
+            except:  # pylint: disable=bare-except
+                pass
+        except:  # pylint: disable=bare-except
+            logger.error(
+                __("Resolwe listener (handle_update[{}]): Error on data save:\n\n{}", data_id, traceback.format_exc()),
+                extra={
+                    'data_id': data_id
+                }
+            )
+
+        if not internal_call:
+            self._send_reply(obj, {ExecutorProtocol.RESULT: ExecutorProtocol.RESULT_OK})
+
+    def handle_finish(self, obj):
+        """Handle an incoming Data finished processing notification.
+
+        Command object format:
+        {
+            'command': 'finish',
+            'id': [id of the Data object this command changes],
+            'process_rc': [exit status of the processing]
+            'spawn_processors': [optional; list of spawn dictionaries],
+            'exported_files_mapper': [if spawn_processors present]
+        }
+
+        :param obj: The channels message object.
+        """
+        data_id = obj[ExecutorProtocol.DATA_ID]
+        str_id = str(data_id)  # json only has string keys, so hydrate_spawned_files will need this stringified
+        logger.debug(
+            __("Resolwe listener: handle_finish[{}].", data_id),
+            extra={
+                'data_id': data_id,
+                'packet': obj
+            }
+        )
+
+        # handle Data wrap up first, so any spawned processors have a consistent view already
+        if ExecutorProtocol.FINISH_PROCESS_RC in obj:
+            process_rc = obj[ExecutorProtocol.FINISH_PROCESS_RC]
+
+            try:
+                d = Data.objects.get(pk=data_id)
+            except Data.DoesNotExist:
+                logger.error(__("Resolwe listener (handle_finish[{}]): Data object does not exist.", data_id))
+                self._send_reply(obj, {ExecutorProtocol.RESULT: ExecutorProtocol.RESULT_OK})
+                return
+
+            if process_rc == 0 and not d.status == Data.STATUS_ERROR:
+                changeset = {
+                    'status': Data.STATUS_DONE,
+                    'process_progress': 100,
+                    'finished': now()
+                }
+            else:
+                changeset = {
+                    'status': Data.STATUS_ERROR,
+                    'process_progress': 100,
+                    'process_rc': process_rc,
+                    'finished': now()
+                }
+            obj[ExecutorProtocol.UPDATE_CHANGESET] = changeset
+            self.handle_update(obj, internal_call=True)
+
+            try:
+                # Cleanup after processor
+                if not getattr(settings, 'FLOW_MANAGER_TEST_KEEP_DATA', False):
```

Comment:
I also don't like this name since we don't want test-specific functionality in the main code.

What about simply using `FLOW_MANAGER_KEEP_DATA`?

Reply:
That doesn't absolve it from being test-specific functionality in the main code. The sole purpose of this variable is to support the functionality of the former keep_data flag in the testcase class. The naming should reflect this specificity.